File: ios/Classes/Renderer/MarkerRenderer.swift
import Foundation
import UIKit

final class MarkerRenderer: ChartLayerRenderer {
  func render(frame: ChartFrame) {
    guard frame.viewport.hasVisibleRange() else { return }

    let visibleMarkers = frame.markerStore.visibleMarkers(
      startTimestamp: frame.candleStore.timestamp(at: frame.viewport.visibleStartIndex),
      endTimestamp: frame.candleStore.timestamp(at: frame.viewport.visibleEndIndex)
    )
    guard !visibleMarkers.isEmpty else { return }

    let context = frame.context
    UIGraphicsPushContext(context)
    defer { UIGraphicsPopContext() }

    for marker in visibleMarkers {
      let candleIndex = frame.candleStore.indexOfTimestamp(marker.timestamp)
      guard candleIndex >= frame.viewport.visibleStartIndex, candleIndex <= frame.viewport.visibleEndIndex else {
        continue
      }

      let x = frame.viewport.xCenter(for: candleIndex, plotLeft: frame.mainChartRect.minX)
      let y = frame.viewport.priceToY(
        marker.price,
        top: frame.mainChartRect.minY,
        height: frame.mainChartRect.height
      )
      guard y >= frame.mainChartRect.minY - 24, y <= frame.mainChartRect.maxY + 24 else {
        continue
      }

      let isBuy = marker.type == "buy"
      let path = UIBezierPath()
      if isBuy {
        path.move(to: CGPoint(x: x, y: y - 10))
        path.addLine(to: CGPoint(x: x - 8, y: y + 4))
        path.addLine(to: CGPoint(x: x + 8, y: y + 4))
      } else {
        path.move(to: CGPoint(x: x, y: y + 10))
        path.addLine(to: CGPoint(x: x - 8, y: y - 4))
        path.addLine(to: CGPoint(x: x + 8, y: y - 4))
      }
      path.close()
      context.setFillColor(isBuy ? frame.theme.buyMarkerColor : frame.theme.sellMarkerColor)
      context.addPath(path.cgPath)
      context.fillPath()

      if let label = marker.label, !label.isEmpty {
        let text = label as NSString
        let textSize = text.size(withAttributes: frame.theme.markerAttributes)
        let labelTop = isBuy ? y + 8 : y - textSize.height - 16
        let labelRect = CGRect(
          x: x - textSize.width / 2 - 6,
          y: labelTop,
          width: textSize.width + 12,
          height: textSize.height + 8
        )
        let path = UIBezierPath(roundedRect: labelRect, cornerRadius: 6)
        context.setFillColor(frame.theme.crosshairLabelBackgroundColor)
        context.addPath(path.cgPath)
        context.fillPath()
        text.draw(
          at: CGPoint(x: labelRect.minX + 6, y: labelRect.minY + 4),
          withAttributes: frame.theme.markerAttributes
        )
      }
    }
  }
}

File: android/src/main/kotlin/com/tradechart/plugin/renderer/MarkerRenderer.kt
package com.tradechart.plugin.renderer

import android.graphics.Path
import android.graphics.RectF
import com.tradechart.plugin.engine.ChartFrame

class MarkerRenderer : ChartLayerRenderer {
    override fun render(frame: ChartFrame) {
        if (!frame.viewport.hasVisibleRange()) {
            return
        }

        val visibleMarkers = frame.markerStore.visibleMarkers(
            startTimestamp = frame.candleStore.timestampAt(frame.viewport.visibleStartIndex),
            endTimestamp = frame.candleStore.timestampAt(frame.viewport.visibleEndIndex),
        )
        if (visibleMarkers.isEmpty()) {
            return
        }

        for (marker in visibleMarkers) {
            val candleIndex = frame.candleStore.indexOfTimestamp(marker.timestamp)
            if (candleIndex !in frame.viewport.visibleStartIndex..frame.viewport.visibleEndIndex) {
                continue
            }
            val x = frame.viewport.xCenterForIndex(candleIndex, frame.mainChartRect.left)
            val y = frame.viewport.priceToY(
                price = marker.price,
                top = frame.mainChartRect.top,
                height = frame.mainChartRect.height(),
            )
            if (y < frame.mainChartRect.top - 24f || y > frame.mainChartRect.bottom + 24f) {
                continue
            }

            val isBuy = marker.type == "buy"
            val trianglePath = Path().apply {
                if (isBuy) {
                    moveTo(x, y - 10f)
                    lineTo(x - 8f, y + 4f)
                    lineTo(x + 8f, y + 4f)
                } else {
                    moveTo(x, y + 10f)
                    lineTo(x - 8f, y - 4f)
                    lineTo(x + 8f, y - 4f)
                }
                close()
            }
            frame.canvas.drawPath(
                trianglePath,
                if (isBuy) frame.theme.buyMarkerPaint else frame.theme.sellMarkerPaint,
            )

            marker.label?.takeIf { it.isNotBlank() }?.let { label ->
                val textWidth = frame.theme.markerTextPaint.measureText(label)
                val labelTop = if (isBuy) y + 8f else y - frame.theme.markerTextPaint.textSize - 16f
                val labelRect = RectF(
                    x - textWidth / 2f - 6f,
                    labelTop,
                    x + textWidth / 2f + 6f,
                    labelTop + frame.theme.markerTextPaint.textSize + 8f,
                )
                frame.canvas.drawRoundRect(
                    labelRect,
                    6f,
                    6f,
                    frame.theme.crosshairLabelBackgroundPaint,
                )
                frame.canvas.drawText(
                    label,
                    labelRect.left + 6f,
                    labelRect.bottom - 6f,
                    frame.theme.markerTextPaint,
                )
            }
        }
    }
}
